### DISCUSSION TITLE
picking Contiguous Cells

### DISCUSSION THREAD

**Discussion by zhang9song** [pyvista/pyvista#4385] [Q&A] [answered]
Hi, all
Does anybody has some good ideas to pick contiguous cells? please review image.
I tried to compute cell normals angles to pick, but got error result.
Attached stl model here.

![20230505085753](https://user-images.githubusercontent.com/24886269/236359653-9f949e1b-46a2-4b61-a0e5-9d5e9621183c.png)
[VESSEL_SEGMENT.zip](https://github.com/pyvista/pyvista/files/11402629/VESSEL_SEGMENT.zip)

**Comment by banesullivan** [comment, marked as answer]
What I didn't realize from the initial screenshot was that there are "sharp" edges delineating the "contiguous" sections. This is emphasized from other view angles:

![screenshot](https://github.com/pyvista/pyvista/assets/22067021/b17da5c1-b458-4d7d-b566-a207ef112532)

We can leverage these "sharp" edges to split the mesh. 

We can find these edges by following https://docs.pyvista.org/version/stable/examples/01-filter/extract-edges.html

```py
import pyvista as pv

mesh = pv.read('VESSEL_SEGMENT.stl')

edges = mesh.extract_feature_edges(20)

pl = pv.Plotter()
pl.add_mesh(mesh, opacity=0.9)
pl.add_mesh(edges, color='red', line_width=5)
pl.show()
```

![screenshot (1)](https://github.com/pyvista/pyvista/assets/22067021/20850aa9-5b88-4bcd-8839-90b70b5367ff)

Then we can use a hidden feature of the `compute_normals()` filter to split the vertices on those feature angles:

```py
split = mesh.compute_normals(
    cell_normals=False,
    split_vertices=True,
    feature_angle=20,
)
```

and with that split mesh, we can use the `split_bodies()` filter to extract each of the split sections of the mesh as their own meshes in a `pyvista.MultiBlock` dataset:

```py
bodies = split.split_bodies()
bodies.plot(multi_colors=True)
```
![screenshot (2)](https://github.com/pyvista/pyvista/assets/22067021/1c25f651-5fee-49c9-a618-f017286e84dd)

```py
# remove the desired section from the multiblock
section = bodies.pop(6)

pl = pv.Plotter(mesh)
pl.add_mesh(bodies, color='blue')
pl.add_mesh(section, color='red')
pl.show()
```
![screenshot (3)](https://github.com/pyvista/pyvista/assets/22067021/71940f3a-c7a5-4256-97b8-cc481b572a37)
